### ISSUE TITLE
Table Insert Performance with HNSW Index

### ISSUE THREAD

**Issue by actonxidian99** [pgvector/pgvector#588]
I am Acton, a database testing engineer. Recently, I have been conducting benchmark tests on pgvector 0.7.0.
The documentation suggests that I can improve index build time by adjusting the parameters max_parallel_maintenance_workers and max_parallel_workers.
However, after changing these parameters, I found that the insert performance on the table with an existing index is not satisfactory.
First, I created a benchmark table and an index:

DROP TABLE IF EXISTS TEST256HNSW;
CREATE TABLE TEST256HNSW(
id BIGINT,
v VECTOR(256)
);

CREATE INDEX ON TEST256HNSW USING hnsw(v vector_l2_ops) WITH(m=16,ef_construction=64);
Then, I used a function to generate random 256-dimensional vector data and inserted 10 million rows of data into it. I did not generate new random data for each insertion but used operators to perform simple arithmetic operations to obtain pseudo-random vector data, minimizing the impact of data generation operations on insert performance. I also manually committed every 100,000 rows.
I recorded the execution time of this function. After the index was created, the insert performance of this function took 5-6 minutes for 100,000 rows and 1 hour for 1 million rows. This does not meet my expectations.
My machine has 440GB of memory and 80 CPU cores. Regarding the main database parameter configuration,
I have set maintenance_work_mem = 80GB, max_parallel_maintenance_workers = 47, shared_buffers = 150GB, and max_parallel_workers = 12.
It is worth mentioning that under the current configuration, inserting a fixed piece of data 1 million times only took 45 seconds.
Is there anything I can do to improve the current performance situation? I apologize for any improper wording due to machine translation. Thank you.

**Comment by jkatz** [comment]
I'm unsure of how you "manually committed" every row when ingesting the data, but I'd make a few recommendations:

1. If you can, use COPY
2. If you're using COPY, use BINARY mode to avoid the translation from binary => text => binary, which adds significant overhead. See here for an example:

https://github.com/erikbern/ann-benchmarks/pull/488

3. If you need to use INSERT and the rows are bulked together, use a bulk INSERY statement (multiple rows in the VALUES section).

4. You can also INSERT (and COPY) concurrently, especially with a target machine with those resources. Here is a test on an older version that shows the benefits of doing that:

https://aws.amazon.com/blogs/database/accelerate-hnsw-indexing-and-searching-with-pgvector-on-amazon-aurora-postgresql-compatible-edition-and-amazon-rds-for-postgresql/

**Comment by ankane** [comment]
Hi @actonxidian99, you'll also want to create any indexes after loading the data for best performance ([docs](https://github.com/pgvector/pgvector?tab=readme-ov-file#loading)).

**Comment by actonxidian99** [comment]
Thank you all @jkatz @ankane for your valuable suggestions for  improvement! First, I used the method of bulk insertion of binary data with COPY, which resulted in a slight performance improvement. Later, I found that creating data first and then building indexes performed 10 to 20 times faster than inserting data first and then building indexes.

Is this performance normal? Inserting data for 1 million rows first and then building indexes only took 2 minutes. However, building indexes first and then inserting data (COPY) took as long as 70 to 80 minutes.
